File: Controllers/EnquiriesController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using PackagingPortal.Api.Data;
using PackagingPortal.Api.DTOs;
using PackagingPortal.Api.Models;

namespace PackagingPortal.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class EnquiriesController : ControllerBase
{
    private readonly AppDbContext _context;
    private readonly ILogger<EnquiriesController> _logger;

    public EnquiriesController(AppDbContext context, ILogger<EnquiriesController> logger)
    {
        _context = context;
        _logger = logger;
    }

    // POST /api/enquiries
    [HttpPost]
    public async Task<ActionResult<EnquiryResponseDto>> Create([FromBody] CreateEnquiryDto dto)
    {
        if (!ModelState.IsValid)
            return BadRequest(ModelState);

        var enquiry = new Enquiry
        {
            Name = dto.Name,
            CompanyName = dto.CompanyName,
            PhoneNumber = dto.PhoneNumber,
            Email = dto.Email,
            Subject = dto.Subject,
            Message = dto.Message,
            Status = "NEW",
            CreatedAt = DateTime.UtcNow
        };

        _context.Enquiries.Add(enquiry);
        await _context.SaveChangesAsync();

        _logger.LogInformation("New enquiry #{Id} from {Name} ({Email})", enquiry.Id, enquiry.Name, enquiry.Email);

        var response = MapToResponse(enquiry);
        return CreatedAtAction(nameof(GetById), new { id = enquiry.Id }, response);
    }

    // GET /api/enquiries
    [HttpGet]
    public async Task<ActionResult<IEnumerable<EnquiryResponseDto>>> GetAll(
        [FromQuery] string? status,
        [FromQuery] string? search,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20)
    {
        var query = _context.Enquiries.AsQueryable();

        if (!string.IsNullOrWhiteSpace(status))
            query = query.Where(e => e.Status == status);

        if (!string.IsNullOrWhiteSpace(search))
            query = query.Where(e => e.Name.Contains(search) || (e.CompanyName != null && e.CompanyName.Contains(search)) || e.Email.Contains(search));

        var total = await query.CountAsync();

        var enquiries = await query
            .OrderByDescending(e => e.CreatedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(e => MapToResponse(e))
            .ToListAsync();

        Response.Headers.Append("X-Total-Count", total.ToString());
        return Ok(enquiries);
    }

    // GET /api/enquiries/{id}
    [HttpGet("{id}")]
    public async Task<ActionResult<EnquiryResponseDto>> GetById(int id)
    {
        var enquiry = await _context.Enquiries.FindAsync(id);
        if (enquiry == null)
            return NotFound();

        return Ok(MapToResponse(enquiry));
    }

    // PUT /api/enquiries/{id}/status
    [HttpPut("{id}/status")]
    public async Task<ActionResult<EnquiryResponseDto>> UpdateStatus(int id, [FromBody] UpdateStatusDto dto)
    {
        var enquiry = await _context.Enquiries.FindAsync(id);
        if (enquiry == null)
            return NotFound();

        enquiry.Status = dto.Status;
        await _context.SaveChangesAsync();

        return Ok(MapToResponse(enquiry));
    }

    // PUT /api/enquiries/{id}/notes
    [HttpPut("{id}/notes")]
    public async Task<ActionResult<EnquiryResponseDto>> UpdateNotes(int id, [FromBody] UpdateNotesDto dto)
    {
        var enquiry = await _context.Enquiries.FindAsync(id);
        if (enquiry == null)
            return NotFound();

        enquiry.Notes = dto.Notes;
        await _context.SaveChangesAsync();

        return Ok(MapToResponse(enquiry));
    }

    private static EnquiryResponseDto MapToResponse(Enquiry e) => new()
    {
        Id = e.Id,
        Name = e.Name,
        CompanyName = e.CompanyName,
        PhoneNumber = e.PhoneNumber,
        Email = e.Email,
        Subject = e.Subject,
        Message = e.Message,
        Status = e.Status,
        Notes = e.Notes,
        CreatedAt = e.CreatedAt
    };
}

File: Spring-boot-backend/package-master-service/src/main/java/com/packagemaster/service/controller/EnquiriesController.java
package com.packagemaster.service.controller;

import com.packagemaster.service.dto.CreateEnquiryDto;
import com.packagemaster.service.dto.EnquiryResponseDto;
import com.packagemaster.service.dto.UpdateNotesDto;
import com.packagemaster.service.dto.UpdateStatusDto;
import com.packagemaster.service.model.Enquiry;
import com.packagemaster.service.repository.EnquiryRepository;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@RestController
@RequestMapping("/api/enquiries")
public class EnquiriesController {

    private final EnquiryRepository enquiryRepository;

    public EnquiriesController(EnquiryRepository enquiryRepository) {
        this.enquiryRepository = enquiryRepository;
    }

    // POST /api/enquiries
    @PostMapping
    public ResponseEntity<EnquiryResponseDto> create(@Valid @RequestBody CreateEnquiryDto dto) {
        Enquiry enquiry = Enquiry.builder()
                .name(dto.getName())
                .companyName(dto.getCompanyName())
                .phoneNumber(dto.getPhoneNumber())
                .email(dto.getEmail())
                .subject(dto.getSubject())
                .message(dto.getMessage())
                .status("NEW")
                .createdAt(Instant.now())
                .build();

        Enquiry saved = enquiryRepository.save(enquiry);
        log.info("New enquiry #{} from {} ({})", saved.getId(), saved.getName(), saved.getEmail());

        EnquiryResponseDto response = mapToResponse(saved);
        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(saved.getId())
                .toUri();

        return ResponseEntity.created(location).body(response);
    }

    // GET /api/enquiries
    @GetMapping
    public ResponseEntity<List<EnquiryResponseDto>> getAll(
            @RequestParam(required = false) String status,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize,
            HttpServletResponse response) {

        Specification<Enquiry> spec = (root, query, cb) -> cb.conjunction();

        if (status != null && !status.trim().isEmpty()) {
            spec = spec.and((root, query, cb) -> cb.equal(root.get("status"), status));
        }

        if (search != null && !search.trim().isEmpty()) {
            String searchPattern = "%" + search.toLowerCase() + "%";
            spec = spec.and((root, query, cb) -> cb.or(
                    cb.like(cb.lower(root.get("name")), searchPattern),
                    cb.like(cb.lower(root.get("companyName")), searchPattern),
                    cb.like(cb.lower(root.get("email")), searchPattern)
            ));
        }

        // Map 1-based page to 0-based page for Spring Data
        int pageIndex = Math.max(0, page - 1);
        Pageable pageable = PageRequest.of(pageIndex, pageSize, Sort.by(Sort.Direction.DESC, "createdAt"));

        Page<Enquiry> enquiryPage = enquiryRepository.findAll(spec, pageable);

        response.setHeader("X-Total-Count", String.valueOf(enquiryPage.getTotalElements()));

        List<EnquiryResponseDto> dtos = enquiryPage.getContent().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());

        return ResponseEntity.ok(dtos);
    }

    // GET /api/enquiries/{id}
    @GetMapping("/{id}")
    public ResponseEntity<EnquiryResponseDto> getById(@PathVariable int id) {
        return enquiryRepository.findById(id)
                .map(this::mapToResponse)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    // PUT /api/enquiries/{id}/status
    @PutMapping("/{id}/status")
    public ResponseEntity<EnquiryResponseDto> updateStatus(
            @PathVariable int id,
            @Valid @RequestBody UpdateStatusDto dto) {

        return enquiryRepository.findById(id)
                .map(enquiry -> {
                    enquiry.setStatus(dto.getStatus());
                    Enquiry saved = enquiryRepository.save(enquiry);
                    return ResponseEntity.ok(mapToResponse(saved));
                })
                .orElse(ResponseEntity.notFound().build());
    }

    // PUT /api/enquiries/{id}/notes
    @PutMapping("/{id}/notes")
    public ResponseEntity<EnquiryResponseDto> updateNotes(
            @PathVariable int id,
            @RequestBody UpdateNotesDto dto) {

        return enquiryRepository.findById(id)
                .map(enquiry -> {
                    enquiry.setNotes(dto.getNotes());
                    Enquiry saved = enquiryRepository.save(enquiry);
                    return ResponseEntity.ok(mapToResponse(saved));
                })
                .orElse(ResponseEntity.notFound().build());
    }

    private EnquiryResponseDto mapToResponse(Enquiry e) {
        return EnquiryResponseDto.builder()
                .id(e.getId())
                .name(e.getName())
                .companyName(e.getCompanyName())
                .phoneNumber(e.getPhoneNumber())
                .email(e.getEmail())
                .subject(e.getSubject())
                .message(e.getMessage())
                .status(e.getStatus())
                .notes(e.getNotes())
                .createdAt(e.getCreatedAt())
                .build();
    }
}
